### ISSUE TITLE
llama_decode return logbits whose value are all nan

### ISSUE THREAD

**Issue by azsane** [ggerganov/llama.cpp#6957]
**enviroment**
gpu : nvidia titan rtx (compute capability 7.5) with 24 GB vram
os : ubuntu 22.04
driver version :550.76
git commit:b4e4b8a9
model:starcode-7B q8_0.v2

**problem encounted**
llama_model ,llama_batch,llama_ctx was all corrrectly initilized
following code behaves different on different gpus
```
llama_batch batch= //... ;  
int ret =llama_decode(ctx,batch);  
float* logits=llama_get_logits(ctx);  
```

1. on a10,it returns the right value
2. on the titan rtx ,all logits are all nan,leading to unusable result and can't engeter the token generation step
3. use the starcoder-1b model ,titan rtx outputs right value
4. using other frameworks,titan rtx can generate valid answer

**question**

1. it seems model > 3 trigger the problem,is it the problem of gpu?
2.is there any way to inspect infer process,seeming that which layer produces nan?

**Comment by slaren** [comment]
> 2.is there any way to inspect infer process,seeming that which layer produces nan?

The [eval-callback example](https://github.com/ggerganov/llama.cpp/tree/master/examples/eval-callback) can help you do that.

**Comment by azsane** [comment]
when callback attached ,i found that serveral tensors are loaded with nan value

for example
![image](https://github.com/ggerganov/llama.cpp/assets/9714904/e7bd2009-6ec4-422c-ab8b-de52475e911f)

**Comment by slaren** [comment]
That's a good start, but we would need know the first operation that produces a `nan`. To do this, you may need to modify the eval-callback example to check for nans on the entire tensor. The `nan` may start as a single isolated `nan` at any location within the tensor, which in the following operations is propagated to the entire activation.

**Comment by SpaceCowboy850** [comment]
I'm not clear where to post this, but this thread has been instrumental in helping me get as far as I've gotten.

So, I'm working on adding a new model to LlamaCPP, and it runs fine with -ngl 31, but -ngl 32 or higher produces garbage output.  It appears the source of the error is coming from the llm_build_kqv method, specifically here:

`        cur = ggml_cont_2d(ctx, kqv_merged, n_embd_head_v*n_head, n_tokens);
        cb(cur, "kqv_merged_cont", il);
`

as kqv_merged_cont-0 is where the first nan's show up.

I'm unclear what ggml_cont_2d, but I'm guessing it's packing memory to be more sequential after a reshape for faster usage.  I can use another model (like, say Phi3 with 4bit quantization, same as ours), and it seems to have no problems.  

We'll continue to investigate, and probably something on our end, but I'm just fishing for other ideas to try.

**Comment by ggerganov** [comment]
Is this with Metal or with CUDA?

**Comment by SpaceCowboy850** [comment]
Cuda

**Comment by SpaceCowboy850** [comment]
Figured it out....comments even stated the problem....reading it's fundamental!

In llm_build_kqv 
This condition needed to be updated with our model

```
        if (model.arch == LLM_ARCH_PHI2 || model.arch == LLM_ARCH_PHI3 || model.arch == LLM_ARCH_GPTNEOX) {
            // for this arch, we need to perform the KQ multiplication with F32 precision, otherwise we get NaNs
            // ref: https://github.com/ggerganov/llama.cpp/pull/4490#issuecomment-1859055847
            ggml_mul_mat_set_prec(kq, GGML_PREC_F32);
        }
```

Which the comment pretty much nails what our problems were, so hat tip to the comment here.
